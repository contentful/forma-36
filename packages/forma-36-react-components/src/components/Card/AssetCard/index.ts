export * from './AssetCard';
export { default } from './AssetCard';
