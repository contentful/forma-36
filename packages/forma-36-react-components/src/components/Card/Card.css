@import 'settings/colors';
@import 'settings/typography';
@import 'settings/transitions';
@import 'settings/borders';
@import 'settings/shadows';

.Card {
  display: block;
  padding: calc(1rem * (14 / var(--font-base-default)));
  border: 1px solid var(--color-element-mid);
  border-radius: var(--border-radius-medium);
  color: var(--color-text-dark);
  text-decoration: none;
  background: var(--color-white);
  box-shadow: var(--box-shadow-default);
  transition: border-color var(--transition-duration-default)
    var(--transition-easing-default);
}

.Card--padding-default {
  padding: calc(1rem * (14 / var(--font-base-default)));
}

.Card--padding-large {
  padding: calc(1rem * (24 / var(--font-base-default)));
}

.Card--padding-none {
  padding: 0;
}

.Card--is-interactive:hover,
.Card--is-interactive:focus {
  outline: none;
  border-color: var(--color-blue-dark);
  cursor: pointer;
}

.Card--is-selected {
  box-shadow: 0px 0px 0px 3px inset var(--color-blue-dark),
    var(--box-shadow-default);
}
