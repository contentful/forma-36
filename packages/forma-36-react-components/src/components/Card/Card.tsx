import React, { ReactNode, MouseEvent, KeyboardEvent } from 'react';
import cn from 'classnames';
import styles from './Card.css';

export interface BaseCardProps {
  /**
   * Used to make the decision of either rendering the card as a link tag or as a div tag
   */
  href?: string;
  /**
   * Used with href to specify target attribute value
   */
  target?: React.AnchorHTMLAttributes<HTMLAnchorElement>['target'];
  /**
   * Applies selected styles to the element
   */
  selected?: boolean;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id)
   */
  testId?: string;
  /**
   * Class names to be appended to the className prop of the component
   */
  className?: string;
  /**
   * Child nodes to be rendered in the component
   */
  children?: ReactNode;
}

export interface CardProps extends BaseCardProps {
  /**
   * The action to be performed when user clicks on the Card
   */
  onClick?: (e: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>) => void;
  /**
   * Applies padding styles of different sizes
   */
  padding?: 'default' | 'large' | 'none';
  /**
   * The title of the entry. It will also be used as aria-label
   */
  title?: string;
  /**
   * Any additional styles that are being applied
   */
  style?: React.CSSProperties;
  /**
   * If the card is selectable and has no title, it will need a aria-label to help screen readers identify it
   */
  ariaLabel?: string;
}

export const Card = ({
  className,
  testId = 'cf-ui-card',
  children,
  href,
  target,
  onClick,
  padding = 'default',
  selected = false,
  ariaLabel,
  ...otherProps
}: CardProps) => {
  const classNames = cn(styles.Card, className, {
    [styles[`Card--padding-${padding}`]]: padding,
    [styles['Card--is-interactive']]: onClick || href,
    [styles['Card--is-selected']]: selected,
  });

  const Element: React.ElementType = href ? 'a' : 'div';

  const handleClick = onClick
    ? (event: MouseEvent<HTMLElement>) => onClick(event)
    : undefined;

  const handleKeyDown = onClick
    ? (event: KeyboardEvent<HTMLElement>) => {
        if (
          event.nativeEvent.code === 'Enter' ||
          event.nativeEvent.code === 'Space'
        ) {
          onClick(event);
          event.currentTarget.focus();
        }
      }
    : undefined;

  return (
    <Element
      className={classNames}
      href={href}
      target={target}
      data-test-id={testId}
      aria-label={otherProps.title || ariaLabel}
      aria-pressed={onClick ? (selected ? 'true' : 'false') : undefined}
      tabIndex={onClick ? 0 : undefined}
      role={onClick ? 'button' : undefined}
      onClick={handleClick}
      onKeyDown={handleKeyDown}
      {...otherProps}
    >
      {children}
    </Element>
  );
};

export default Card;
