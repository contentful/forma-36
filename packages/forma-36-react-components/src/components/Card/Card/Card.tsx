import React, { Component, SyntheticEvent } from 'react';
import cn from 'classnames';
const styles = require('./Card.css');

export interface CardPropTypes {
  extraClassNames?: string;
  children: React.ReactNode;
  href?: string;
  onClick?: (e: SyntheticEvent) => void;
  testId?: string;
  padding?: 'default' | 'large' | 'none';
  selected?: boolean;
  title?: string;
  style?: React.CSSProperties;
}

export class Card extends Component<CardPropTypes> {
  static defaultProps = {
    extraClassNames: undefined,
    href: undefined,
    onClick: undefined,
    padding: 'default',
    testId: 'cf-ui-card',
    selected: false,
  };

  handleClick = event => {
    if (!this.props.onClick) return;

    this.props.onClick(event);
  };

  render() {
    const {
      extraClassNames,
      testId,
      children,
      href,
      onClick,
      padding,
      selected,
      ...otherProps
    } = this.props;

    const classNames = cn(styles.Card, extraClassNames, {
      [styles[`Card--padding-${padding}`]]: padding,
      [styles['Card--is-interactive']]: onClick || href,
      [styles['Card--is-selected']]: selected,
    });

    const Element: React.ReactType = href ? 'a' : 'div';
    return React.createElement(
      Element,
      {
        href,
        className: classNames,
        onClick: this.handleClick,
        'data-test-id': testId,
        ...otherProps,
      },
      children,
    );
  }
}

export default Card;
