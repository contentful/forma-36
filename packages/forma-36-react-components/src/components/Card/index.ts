export * from './Card';
export { default } from './Card';
