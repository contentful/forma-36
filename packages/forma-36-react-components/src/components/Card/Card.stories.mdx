import { useState } from 'react';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';

import { Typography } from '../Typography/Typography';
import { Heading } from '../Typography/Heading';
import { Paragraph } from '../Typography/Paragraph';

import { Card } from './Card';

<Meta title="Components/Card" component={Card} />

# Card

[source code](https://github.com/contentful/forma-36/blob/master/packages/forma-36-react-components/src/components/Card/Card.tsx)

A Card is a component to group information or content of a particular subject in one place.

<Canvas>
  <Story
    name="default"
    args={{ title: 'Title', content: 'This is the Card‚Äôs content' }}
  >
    {({ title, content, ...args }) => (
      <Card {...args}>
        <Typography>
          <Heading>{title}</Heading>
          <Paragraph>{content}</Paragraph>
        </Typography>
      </Card>
    )}
  </Story>
</Canvas>

## Table of contents

- [Examples of usage](#examples-of-usage)
  - [With onClick](#with-onclick)
  - [Selectable cards](#selectable-cards)
  - [With link and target](#with-link-and-target)
- [Props](#props)

## Examples of usage

Here you will find the most common ways of using the Card component.

### With `onClick`

It‚Äôs possible to call an action when the user clicks on the Card. To do that, you only need to pass a function to the `onClick` prop.

<Canvas>
  <Story name="with onClick">
    {(args) => {
      const [show, setShow] = useState(false);
      return (
        <div>
          <Card
            {...args}
            onClick={() => setShow(!show)}
            style={{ textAlign: 'center' }}
          >
            Click on this card
            <br />
            {show && (
              <span role="img" aria-label="sparkles">
                ‚ú®‚ú®‚ú®
              </span>
            )}
          </Card>
        </div>
      );
    }}
  </Story>
</Canvas>

### Selectable cards

Sometimes you will need to use the Card as an selectable element in the UI. To achieve that behaviour,
you should use a combination of the `onClick` and `selected` props. The former controlling the latter.

<Canvas>
  <Story name="selectable cards">
    {(args) => {
      const [taco, setTaco] = useState(false);
      const [pizza, setPizza] = useState(false);
      const [broccoli, setBroccoli] = useState(false);
      return (
        <div style={{ maxWidth: '280px' }}>
          <Heading>What is your favorite food?</Heading>
          <div
            style={{
              display: 'flex',
              justifyContent: 'space-between',
              marginTop: '1rem',
            }}
          >
            <Card onClick={() => setTaco(!taco)} selected={taco}>
              <span style={{ fontSize: '3rem' }} role="img" aria-label="taco">
                üåÆ
              </span>
            </Card>
            <Card onClick={() => setPizza(!pizza)} selected={pizza}>
              <span style={{ fontSize: '3rem' }} role="img" aria-label="pizza">
                üçï
              </span>
            </Card>
            <Card onClick={() => setBroccoli(!broccoli)} selected={broccoli}>
              <span
                style={{ fontSize: '3rem' }}
                role="img"
                aria-label="broccoli"
              >
                ü•¶
              </span>
            </Card>
          </div>
        </div>
      );
    }}
  </Story>
</Canvas>

### With `link` and `target`

A Card can be used with the pourpose to navigate a user to an external page. To do that a user must pass a value to the `href` prop.
Optionally, the user can pass a `target` prop to control how the link should be open.

<Canvas>
  <Story
    name="with link and target"
    args={{ href: 'http://f36.contentful.com', target: '_blank' }}
  >
    {(args) => (
      <Card {...args}>
        <Typography>
          <Heading>Forma36</Heading>
          <Paragraph>
            Forma 36 is an open-source design system by Contentful created with
            the intent to reduce the overhead of creating UI by providing tools
            and guidance for digital teams building and extending Contentful
            products.
          </Paragraph>
        </Typography>
      </Card>
    )}
  </Story>
</Canvas>

## Props

<ArgsTable of={Card} />
