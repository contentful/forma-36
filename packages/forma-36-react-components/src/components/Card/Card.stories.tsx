import React, { useState } from 'react';

import { Typography } from '../Typography/Typography';
import { Heading } from '../Typography/Heading';
import { Paragraph } from '../Typography/Paragraph';

import { Card, CardProps } from './Card';

export default {
  title: 'Components/Card',
  component: Card,
  parameters: {
    propTypes: Card['__docgenInfo'],
  },
};

export const Default = ({ children, ...args }: CardProps) => {
  return (
    <Card {...args}>
      <Typography>
        <Paragraph>{children}</Paragraph>
      </Typography>
    </Card>
  );
};
Default.args = { title: 'Title', children: 'This is the Card‚Äôs content' };

export const WithOnClick = (args: CardProps) => {
  const [show, setShow] = useState(false);
  return (
    <div>
      <Card
        {...args}
        onClick={() => setShow(!show)}
        style={{ textAlign: 'center' }}
      >
        Click on this card
        <br />
        {show && (
          <span role="img" aria-label="sparkles">
            ‚ú®‚ú®‚ú®
          </span>
        )}
      </Card>
    </div>
  );
};

export const SelectableCards = () => {
  const [taco, setTaco] = useState(false);
  const [pizza, setPizza] = useState(false);
  const [broccoli, setBroccoli] = useState(false);
  return (
    <div style={{ maxWidth: '280px' }}>
      <Heading>What is your favorite food?</Heading>
      <div
        style={{
          display: 'flex',
          justifyContent: 'space-between',
          marginTop: '1rem',
        }}
      >
        <Card onClick={() => setTaco(!taco)} selected={taco}>
          <span style={{ fontSize: '3rem' }} role="img" aria-label="taco">
            üåÆ
          </span>
        </Card>
        <Card onClick={() => setPizza(!pizza)} selected={pizza}>
          <span style={{ fontSize: '3rem' }} role="img" aria-label="pizza">
            üçï
          </span>
        </Card>
        <Card onClick={() => setBroccoli(!broccoli)} selected={broccoli}>
          <span style={{ fontSize: '3rem' }} role="img" aria-label="broccoli">
            ü•¶
          </span>
        </Card>
      </div>
    </div>
  );
};

export const WithLinkAndTarget = (args: CardProps) => {
  return (
    <Card {...args}>
      <Typography>
        <Heading>Forma36</Heading>
        <Paragraph>
          Forma 36 is an open-source design system by Contentful created with
          the intent to reduce the overhead of creating UI by providing tools
          and guidance for digital teams building and extending Contentful
          products.
        </Paragraph>
      </Typography>
    </Card>
  );
};
WithLinkAndTarget.args = {
  href: 'https://f36.contentful.com/',
  target: '_blank',
};
