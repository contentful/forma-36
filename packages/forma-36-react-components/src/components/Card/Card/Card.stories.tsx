import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, boolean, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

import Card from './Card';

storiesOf('Components|Card', module)
  .addParameters({
    propTypes: Card['__docgenInfo'],
  })
  .add('default', () => (
    <Card
      className={text('className', '')}
      href={text('href', '') || undefined}
      padding={select(
        'padding',
        {
          Default: 'default',
          Large: 'large',
          None: 'none',
        },
        'default',
      )}
      selected={boolean('selected', false)}
    >
      {text('children', 'Card')}
    </Card>
  ))
  .add('with onClick handler', () => (
    <Card
      className={text('className', '')}
      href={text('href', '') || undefined}
      onClick={action('onClick')}
      padding={select(
        'padding',
        {
          Default: 'default',
          Large: 'large',
          None: 'none',
        },
        'default',
      )}
      selected={boolean('selected', false)}
    >
      {text('children', 'Card')}
    </Card>
  ));
