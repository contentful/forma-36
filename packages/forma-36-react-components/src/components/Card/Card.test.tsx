import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from '@/scripts/test/axeHelper';

import { Card } from './Card';

it('renders the component', () => {
  const { container } = render(<Card>Card</Card>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(<Card className="my-extra-class">Card</Card>);

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<Card>Card</Card>);
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});

it('renders as a link if passed a href prop', () => {
  const { container } = render(<Card href="/">Card</Card>);

  expect(container.firstChild).toMatchSnapshot();
});

it('has an "is-interactive" class if has an onClick prop', () => {
  const { container } = render(<Card href="/">Card</Card>);

  expect(container.firstChild).toMatchSnapshot();
});

it('calls an onClick function', () => {
  const onClickFunc = jest.fn();
  const { container } = render(<Card onClick={onClickFunc}>Card</Card>);
  const element = container.firstChild as Element;
  userEvent.click(element);

  expect(container.firstChild).toMatchSnapshot();
  expect(onClickFunc).toHaveBeenCalled();
});

it('can be selected', () => {
  const output = render(<Card selected>Card</Card>);

  expect(output).toMatchSnapshot();
});
