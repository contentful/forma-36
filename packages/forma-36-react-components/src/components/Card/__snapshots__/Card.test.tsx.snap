// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calls an onClick function 1`] = `
<button
  aria-pressed="false"
  class="Card Card--padding-default Card--is-interactive"
  data-test-id="cf-ui-card"
>
  Card
</button>
`;

exports[`can be selected 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="Card Card--padding-default Card--is-selected"
        data-test-id="cf-ui-card"
      >
        Card
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="Card Card--padding-default Card--is-selected"
      data-test-id="cf-ui-card"
    >
      Card
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`has an "is-interactive" class if has an onClick prop 1`] = `
<a
  class="Card Card--padding-default Card--is-interactive"
  data-test-id="cf-ui-card"
  href="/"
>
  Card
</a>
`;

exports[`renders as a link if passed a href prop 1`] = `
<a
  class="Card Card--padding-default Card--is-interactive"
  data-test-id="cf-ui-card"
  href="/"
>
  Card
</a>
`;

exports[`renders the component 1`] = `
<div
  class="Card Card--padding-default"
  data-test-id="cf-ui-card"
>
  Card
</div>
`;

exports[`renders the component with an additional class name 1`] = `
<div
  class="Card my-extra-class Card--padding-default"
  data-test-id="cf-ui-card"
>
  Card
</div>
`;
